File: java/src/main/java/gherkin/Token.java
package gherkin;

import gherkin.ast.Location;

import java.util.List;

public class Token {
    public boolean IsEOF() {
        return line == null;
    }

    public final IGherkinLine line;
    public Parser.TokenType matchedType;
    public String matchedKeyword;
    public String matchedText;
    public List<GherkinLineSpan> mathcedItems;
    public int matchedIndent;
    public GherkinDialect matchedGherkinDialect;
    public Location location;

    public Token(IGherkinLine line, Location location) {
        this.line = line;
        this.location = location;
    }

    public void detach() {
        if (line != null)
            line.detach();
    }

    public String getTokenValue() {
        return IsEOF() ? "EOF" : line.getLineText(-1);
    }

    @Override
    public String toString() {
        return String.format("%s: %s/%s", matchedType, matchedKeyword, matchedText);
    }

}

File: elixir/lib/gherkin/token.ex
defmodule Gherkin.Token do

  alias Gherkin.GherkinLine

  defstruct line: nil,
            location: nil,
            matched_type: nil,
            matched_text: nil,
            matched_keyword: nil,
            matched_indent: nil,
            matched_items: nil,
            matched_gherkin_dialect: nil

  def new(location) do
    %__MODULE__{location: location}
  end

  def new(line, location) do
    %__MODULE__{line: line, location: location}
  end

  def eof?(%{line: nil}), do: true
  def eof?(_), do: false

  def detach do
    # TODO: detach line - is this needed?
  end

  def token_value(%{line: nil}), do: "EOF"
  def token_value(%{line: line}), do: GherkinLine.get_line_text(line, -1)

end
